### ISSUE TITLE
Prediction and Column Order

### ISSUE THREAD

**Issue by pmgblz** [grf-labs/grf#345]
Predictions from grf seem to change if the column order of the original data is changed. See below for a toy example using regression_forest.

An example of when column order might change is if a user wants to vary a variable of interest, holding others constant, to visualize heterogeneity. In this case, the modified variable of interest is most likely appended to the data and not put back into the dataset at the original column index.

However, the column order of the training data should generally be stable. Therefore, a suggestion would be to prompt an error if the user passes new data with different columns from the training set.

```
library(grf)

# generate data
set.seed(123)
x <- data.frame(a = rnorm(100), 
                b = rnorm(100), 
                c = rnorm(100), 
                d = rnorm(100))

y <- rnorm(100)

# reshuffle columns in x
x_reshuffle <- x[ , c("b", "a", "d", "c")]

# run regression forest
gfr_rf <- grf::regression_forest(x, y)

# get predictions for original and reshuffled data
prediction_x <- predict(gfr_rf, newdata=x)$predictions # predict on x
prediction_x_reshuffle <- predict(gfr_rf, newdata=x_reshuffle)$predictions # predict on reshuffled x

# predicting on original and reshuffled data does not give the same predictions
# note: also seems to happen when x is a matrix and not a data frame
summary(prediction_x) == summary(prediction_x_reshuffle)
```

**Comment by jtibshirani** [comment]
Hi @pmgblz, thanks for the issue and apologies that this caused confusion.

I don't think it's too uncommon for packages to expect the test data to contain the same columns (in the same order) as the test data (for example, `glmnet` has this requirement). To me it feels a little heavy-handed to check that the training and test set have the same column names, as this doesn't always indicate something concerning or obviously wrong.

For now I've updated the documentation in #348 to try to avoid this confusion in the future. I'll also plan to submit a PR that verifies the training and test data have the same number of columns.
